File: libs/daemon/client/src/mill/client/FileToStreamTailer.java
package mill.client;

import java.io.BufferedReader;
import java.io.File;
import java.io.IOException;
import java.io.PrintStream;
import java.nio.file.Files;
import java.util.Optional;

public class FileToStreamTailer extends Thread implements AutoCloseable {

  private final File file;
  private final PrintStream stream;
  private final int intervalMsec;

  // if true, we won't read the whole file, but only new lines
  private boolean ignoreHead = true;

  private volatile boolean keepReading = true;
  private volatile boolean flush = false;

  public FileToStreamTailer(File file, PrintStream stream, int intervalMsec) {
    super("FileToStreamTailerThread");
    this.intervalMsec = intervalMsec;
    setDaemon(true);
    this.file = file;
    this.stream = stream;
  }

  @Override
  public void run() {
    if (isInterrupted()) {
      keepReading = false;
    }
    Optional<BufferedReader> reader = Optional.empty();
    try {
      while (keepReading || flush) {
        flush = false;
        try {
          // Init reader, if not already done
          if (!reader.isPresent()) {
            try {
              reader = Optional.of(Files.newBufferedReader(file.toPath()));
            } catch (IOException e) {
              // nothing to ignore if file is initially missing
              ignoreHead = false;
            }
          }
          reader.ifPresent(r -> {
            // read lines
            try {
              String line;
              while ((line = r.readLine()) != null) {
                if (!ignoreHead) {
                  stream.println(line);
                }
              }
              // we ignored once
              this.ignoreHead = false;
            } catch (IOException e) {
              // could not read line or file vanished
            }
          });
        } finally {
          if (keepReading) {
            // wait
            try {
              Thread.sleep(intervalMsec);
            } catch (InterruptedException e) {
              // can't handle anyway
            }
          }
        }
      }
    } finally {
      reader.ifPresent(r -> {
        try {
          r.close();
        } catch (IOException e) {
          // could not close but also can't do anything about it
        }
      });
    }
  }

  @Override
  public void interrupt() {
    this.keepReading = false;
    super.interrupt();
  }

  /**
   * Force a next read, even if we interrupt the thread.
   */
  public void flush() {
    this.flush = true;
  }

  @Override
  public void close() throws Exception {
    flush();
    interrupt();
  }
}

File: libs/daemon/client/src/mill/client/FileToStreamTailer.scala
package mill.client

import java.io.{BufferedReader, InputStreamReader, PrintStream}

class FileToStreamTailer(file: os.Path, stream: PrintStream, intervalMsec: Int)
    extends Thread("FileToStreamTailerThread") with AutoCloseable {

  setDaemon(true)

  // if true, we won't read the whole file, but only new lines
  @volatile private var ignoreHead = true
  @volatile private var keepReading = true
  @volatile private var doFlush = false

  override def run(): Unit = {
    if (isInterrupted) keepReading = false

    var reader: BufferedReader = null
    try {
      while (keepReading || doFlush) {
        doFlush = false
        try {
          // Init reader, if not already done
          if (reader == null) {
            try reader = new BufferedReader(new InputStreamReader(os.read.inputStream(file)))
            catch {
              case _: java.io.IOException =>
                // nothing to ignore if file is initially missing
                ignoreHead = false
            }
          }
          if (reader != null) { // read lines
            try {
              var line = reader.readLine()
              while (line != null) {
                if (!ignoreHead) stream.println(line)
                line = reader.readLine()
              }

              this.ignoreHead = false // we ignored once
            } catch { case _: java.io.IOException =>  /* could not read line or file vanished*/ }
          }
        } finally {
          if (keepReading) {
            try Thread.sleep(intervalMsec)
            catch { case _: InterruptedException => }
          }
        }
      }
    } finally {
      if (reader != null) {
        try reader.close()
        catch { case _: java.io.IOException => }
      }
    }
  }

  override def interrupt(): Unit = {
    this.keepReading = false
    super.interrupt()
  }

  /** Force a next read, even if we interrupt the thread. */
  def flush(): Unit = {
    this.doFlush = true
  }

  override def close(): Unit = {
    flush()
    interrupt()
  }
}
